mlos_bench/mlos_bench/services/remote/ssh/ssh_service.py

```diff
@@ -0,0 +1,391 @@
+            if client:
+                _LOG.debug("%s: Checking cached client %s", current_thread().name, connection_id)
+                connection = await client.connection()
+                if not connection:
+                    _LOG.debug("%s: Removing stale client connection %s from cache.", current_thread().name, connection_id)
+                    self._cache.pop(connection_id)
+                    # Try to reconnect next.
+                else:
+                    _LOG.debug("%s: Using cached client %s", current_thread().name, connection_id)
+            if connection_id not in self._cache:
+                _LOG.debug("%s: Establishing client connection to %s", current_thread().name, connection_id)
+                connection, client = await asyncssh.create_connection(SshClient, **connect_params)
+                assert isinstance(client, SshClient)
+                self._cache[connection_id] = (connection, client)
+                _LOG.debug("%s: Created connection to %s.", current_thread().name, connection_id)
+            return self._cache[connection_id]
+
+    def cleanup(self) -> None:
+        """
+        Closes all cached connections.
+        """
+        for (connection, _) in self._cache.values():
+            connection.close()
+        self._cache = {}
+
+
+class SshService(Service, metaclass=ABCMeta):
+    """
+    Base class for SSH services.
+    """
+
+    # AsyncSSH requires an asyncio event loop to be running to work.
+    # However, that that event loop blocks the main thread.
+    # To avoid having to change our entire API to use async/await, all the way
+    # up the stack, we run the event loop that runs any async code in a
+    # background thread and submit async code to it using
+    # asyncio.run_coroutine_threadsafe, interacting with Futures after that.
+    # This is a bit of a hack, but it works for now.
+    #
+    # The event loop is created on demand and shared across all SshService
+    # instances, hence we need to lock it when doing the creation/cleanup.
+    #
+    # We ran tests to ensure that multiple requests can still be executing
+    # concurrently inside that event loop so there should be no practical
+    # performance loss for our initial cases even with just single background
+    # thread running the event loop.
+    #
+    # Note: the tests were run to confirm that this works with two threads.
+    # Using a larger thread pool requires a bit more work since asyncssh
+    # requires that run() requests are submitted to the same event loop handler
+    # that the connection was made on.
+    # In that case, each background thread should get its own SshClientCache.
+
+    _event_loop: Optional[AbstractEventLoop] = None
+    _event_loop_thread: Optional[Thread] = None
+    _event_loop_thread_lock = ThreadLock()
+    _event_loop_thread_refcnt: int = 0
+    _event_loop_thread_ssh_client_cache: Optional[SshClientCache] = None
```

Comment:
or maybe even roll all these variables into a single `@dataclass` struct with enter/exit methods so we can write e.g., `SshService._CONTEXT.enter(...)` or `SshService._CONTEXT.clear_cache()`or access the stuff in it like `SshService._CONTEXT.event_loop`

Reply:
I split it out a little bit differently so that it hopefully supports other future uses as well (e.g., parallel trial runners perhaps).
It keeps the event loop manager separately from the client cache, since the latter is somewhat specific to the SshService.
Let me know what you think: #555